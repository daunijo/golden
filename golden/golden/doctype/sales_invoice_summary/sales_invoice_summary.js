// Copyright (c) 2018, RSS and contributors
// For license information, please see license.txt

frappe.ui.form.on('Sales Invoice Summary', {
	refresh: function(frm) {
		frm.events.set_read_only(frm);
	},
	set_posting_time: function(frm){
		frm.events.set_read_only(frm);
	},
	set_read_only: function(frm){
		if(frm.doc.set_posting_time == 1){
			frm.set_df_property("posting_date", "read_only", false);
			frm.set_df_property("posting_time", "read_only", false);
		}else{
			frm.set_df_property("posting_date", "read_only", true);
			frm.set_df_property("posting_time", "read_only", true);
		}
	},
	get_sales_invoice: function(frm){
		return frappe.call({
			method: 'golden.golden.doctype.sales_invoice_summary.sales_invoice_summary.get_sales_invoice',
			args: {
				start: frm.doc.start_date,
				end: frm.doc.end_date,
				sales_person: frm.doc.sales || undefined
			},
			callback: function(r, rt) {
				if(r.message) {
					frm.clear_table("invoices");
					$.each(r.message, function(i, d) {
						var c = frm.add_child("invoices");
						c.customer = d.customer;
						c.customer_name = d.customer_name;
						c.sales_invoice = d.si_name;
						c.invoice_date = d.posting_date;
						c.amount = d.amount;
						c.due_date = d.due_date;
					})
					frm.refresh_fields();
					frm.events.set_total_invoice(frm);
				}
			}
		})
	},
	set_total_invoice: function(frm) {
		var total_inv = 0.0;
		$.each(frm.doc.invoices, function(i, row) {
			total_inv += flt(row.amount);
		})
		frm.set_value("total_invoice", Math.abs(total_inv));
	},
});
frappe.ui.form.on('Sales Invoice Summary Detail', {
	invoices_add: function(frm, cdt, cdn) {
		calculate_total_invoice(frm, cdt, cdn);
	},
	invoices_remove: function(frm, cdt, cdn) {
		calculate_total_invoice(frm, cdt, cdn);
	},
})
var calculate_total_invoice = function(frm) {
	var total_invoice = frappe.utils.sum(
		(frm.doc.invoices || []).map(function(i) {
			return (flt(i.amount));
		})
	);
	frm.set_value("total_invoice", total_invoice);
}
