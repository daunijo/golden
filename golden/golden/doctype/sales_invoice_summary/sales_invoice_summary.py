# -*- coding: utf-8 -*-
# Copyright (c) 2018, RSS and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.model.document import Document
from frappe.utils import nowdate, cstr, flt, now, getdate, add_months
from frappe import msgprint, _
from frappe.model.mapper import get_mapped_doc

class SalesInvoiceSummary(Document):
	def validate(self):
		frappe.db.set(self, 'status', 'Draft')
		self.check_is_invoice_summary()

	def check_is_invoice_summary(self):
		for row in self.invoices:
			if row.reference_doctype == "Sales Invoice":
				check_si = frappe.db.sql("""select invoice_keeptrack from `tabSales Invoice` where docstatus = '1' and `name` = %s""", row.reference_name)[0][0]
				if check_si:
					frappe.throw(_("Invoice {0} already used in Invoice Keeptrack").format(row.reference_name))
			elif row.reference_doctype == "Sales Return":
				check_si = frappe.db.sql("""select sales_invoice_summary from `tabSales Return` where docstatus = '1' and `name` = %s""", row.reference_name)[0][0]
				if check_si:
					frappe.throw(_("Sales Return {0} already used in other Sales Invoice Summary").format(row.reference_name))

	def on_submit(self):
		frappe.db.set(self, 'status', 'Submitted')
		for row in self.invoices:
			if row.reference_doctype == "Sales Invoice":
				frappe.db.sql("""update `tabSales Invoice` set sales_invoice_summary = %s where `name` = %s""", (self.name, row.reference_name))
			elif row.reference_doctype == "Sales Return":
				frappe.db.sql("""update `tabSales Return` set sales_invoice_summary = %s where `name` = %s""", (self.name, row.reference_name))

	def on_cancel(self):
		frappe.db.set(self, 'status', 'Cancelled')
		for row in self.invoices:
			if row.reference_doctype == "Sales Invoice":
				frappe.db.sql("""update `tabSales Invoice` set sales_invoice_summary = null where `name` = %s""", row.reference_name)
			elif row.reference_doctype == "Sales Return":
				frappe.db.sql("""update `tabSales Return` set sales_invoice_summary = null where `name` = %s""", row.reference_name)

@frappe.whitelist()
def get_sales_invoice(customer, start, end, sales):
	si_list = []
	conditions = ""
	conditions += " and si.customer = '%s'" % frappe.db.escape(customer)
	conditions += " and si.posting_date >= '%s'" % frappe.db.escape(start)
	conditions += " and si.posting_date <= '%s'" % frappe.db.escape(end)
	if sales != "":
		conditions += " and si.rss_sales_person = '%s'" % frappe.db.escape(sales)

	invoice_list = frappe.db.sql("""select si.`name` as si_name, si.customer, si.customer_name, si.posting_date, si.grand_total, si.due_date, si.rss_sales_name from `tabSales Invoice` si inner join `tabCustomer` c on c.`name` = si.customer where si.docstatus = '1' and si.`status` != 'Paid' and si.sales_invoice_summary is null %s""" % conditions, as_dict=True)
	for d in invoice_list:
		# count_payment = frappe.db.sql("""select count(*) from `tabPayment Entry Reference` where docstatus = '1' and reference_name = %s""", d.name)[0][0]
		si_list.append(frappe._dict({
			'customer': d.customer,
			'customer_name': d.customer_name,
			'reference_doctype': "Sales Invoice",
			'reference_name': d.si_name,
			'posting_date': d.posting_date,
			'amount': d.grand_total,
			'due_date': d.due_date,
			'sales_name': d.rss_sales_name
		}))
	conditions2 = ""
	conditions2 += " and si.customer = '%s'" % frappe.db.escape(customer)
	conditions2 += " and si.posting_date >= '%s'" % frappe.db.escape(start)
	conditions2 += " and si.posting_date <= '%s'" % frappe.db.escape(end)
	if sales != "":
		conditions2 += " and si.sales_person = '%s'" % frappe.db.escape(sales)
	return_list = frappe.db.sql("""select si.`name` as si_name, si.customer, si.customer_name, si.posting_date, si.total_2 from `tabSales Return` si inner join `tabCustomer` c on c.`name` = si.customer where si.docstatus = '1' and si.sales_invoice_summary is null %s""" % conditions2, as_dict=True)
	for e in return_list:
		si_list.append(frappe._dict({
			'customer': e.customer,
			'customer_name': e.customer_name,
			'reference_doctype': "Sales Return",
			'reference_name': e.si_name,
			'posting_date': e.posting_date,
			'amount': e.total_2,
			'due_date': "",
			'sales_name': ""
		}))
	return si_list
