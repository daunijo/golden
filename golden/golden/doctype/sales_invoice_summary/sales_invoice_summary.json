{
 "allow_copy": 0, 
 "allow_guest_to_view": 0, 
 "allow_import": 0, 
 "allow_rename": 0, 
 "autoname": "naming_series:", 
 "beta": 0, 
 "creation": "2018-02-07 14:39:50.625039", 
 "custom": 0, 
 "docstatus": 0, 
 "doctype": "DocType", 
 "document_type": "", 
 "editable_grid": 1, 
 "engine": "InnoDB", 
 "fields": [
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sb01", 
   "fieldtype": "Section Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "naming_series", 
   "fieldtype": "Select", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Series", 
   "length": 0, 
   "no_copy": 1, 
   "options": "SS-", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 1, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "customer", 
   "fieldtype": "Link", 
   "hidden": 1, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Customer", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Customer", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "depends_on": "customer", 
   "fieldname": "customer_name", 
   "fieldtype": "Data", 
   "hidden": 1, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Customer Name", 
   "length": 0, 
   "no_copy": 0, 
   "options": "customer.customer_name", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "start_date", 
   "fieldtype": "Date", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Start Date", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "end_date", 
   "fieldtype": "Date", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "End Date", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sales", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Sales", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Employee", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sales_name", 
   "fieldtype": "Data", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Sales Name", 
   "length": 0, 
   "no_copy": 0, 
   "options": "sales.employee_name", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "region", 
   "fieldtype": "Data", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Region", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "cb01", 
   "fieldtype": "Column Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "default": "Today", 
   "fieldname": "posting_date", 
   "fieldtype": "Date", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Date", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "posting_time", 
   "fieldtype": "Time", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Posting Time", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "depends_on": "eval:doc.docstatus==0", 
   "fieldname": "set_posting_time", 
   "fieldtype": "Check", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Edit Posting Date and Time", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "company", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 1, 
   "in_standard_filter": 0, 
   "label": "Company", 
   "length": 0, 
   "no_copy": 0, 
   "options": "Company", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 1, 
   "report_hide": 0, 
   "reqd": 1, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "amended_from", 
   "fieldtype": "Link", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Amended From", 
   "length": 0, 
   "no_copy": 1, 
   "options": "Sales Invoice Summary", 
   "permlevel": 0, 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sb02", 
   "fieldtype": "Section Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "depends_on": "eval:doc.start_date && doc.end_date && doc.sales", 
   "fieldname": "get_sales_invoice", 
   "fieldtype": "Button", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Get Sales Invoice", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "invoices", 
   "fieldtype": "Table", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Invoices", 
   "length": 0, 
   "no_copy": 1, 
   "options": "Sales Invoice Summary Detail", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "sb03", 
   "fieldtype": "Section Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "default": "Draft", 
   "fieldname": "status", 
   "fieldtype": "Select", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Status", 
   "length": 0, 
   "no_copy": 1, 
   "options": "\nDraft\nSubmitted\nCompleted\nCancelled", 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 1, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "invoice_keeptrack", 
   "fieldtype": "Data", 
   "hidden": 1, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Invoice Keeptrack", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "cb03", 
   "fieldtype": "Column Break", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 0, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }, 
  {
   "allow_bulk_edit": 0, 
   "allow_on_submit": 0, 
   "bold": 0, 
   "collapsible": 0, 
   "columns": 0, 
   "fieldname": "total_invoice", 
   "fieldtype": "Currency", 
   "hidden": 0, 
   "ignore_user_permissions": 0, 
   "ignore_xss_filter": 0, 
   "in_filter": 0, 
   "in_global_search": 0, 
   "in_list_view": 0, 
   "in_standard_filter": 0, 
   "label": "Total Invoice", 
   "length": 0, 
   "no_copy": 0, 
   "permlevel": 0, 
   "precision": "", 
   "print_hide": 0, 
   "print_hide_if_no_value": 0, 
   "read_only": 1, 
   "remember_last_selected_value": 0, 
   "report_hide": 0, 
   "reqd": 0, 
   "search_index": 0, 
   "set_only_once": 0, 
   "unique": 0
  }
 ], 
 "has_web_view": 0, 
 "hide_heading": 0, 
 "hide_toolbar": 0, 
 "idx": 0, 
 "image_view": 0, 
 "in_create": 0, 
 "is_submittable": 1, 
 "issingle": 0, 
 "istable": 0, 
 "max_attachments": 0, 
 "modified": "2018-07-16 15:16:18.469081", 
 "modified_by": "Administrator", 
 "module": "Golden", 
 "name": "Sales Invoice Summary", 
 "name_case": "", 
 "owner": "Administrator", 
 "permissions": [
  {
   "amend": 1, 
   "apply_user_permissions": 0, 
   "cancel": 1, 
   "create": 1, 
   "delete": 1, 
   "email": 1, 
   "export": 1, 
   "if_owner": 0, 
   "import": 0, 
   "permlevel": 0, 
   "print": 1, 
   "read": 1, 
   "report": 1, 
   "role": "System Manager", 
   "set_user_permissions": 0, 
   "share": 1, 
   "submit": 1, 
   "write": 1
  }
 ], 
 "quick_entry": 0, 
 "read_only": 0, 
 "read_only_onload": 0, 
 "show_name_in_global_search": 0, 
 "sort_field": "modified", 
 "sort_order": "DESC", 
 "track_changes": 1, 
 "track_seen": 0
}